feat(extras/scripts): update Transmission peer port

juanlufont opened the pull request:
The script `transmission-port-update.sh` updates the _peer port_ used by the torrent client Transmission using its API. The peer port is the port used and announced by Transmission to accept incoming connections. You can use this script to set a VPN forwarded port as the current listening port for Transmission.

The script uses only shell scripting and can run out of the box within the current `gluetun` docker image, it only relies on the command `wget`, which is already included by default in the image.

juanlufont committed:
Add script to update Transmission peer port

The script `transmission-port-update.sh` updates the _peer port_ used by
the torrent client Transmission. The peer port is the port used and
announced by Transmission to accept incoming connections.

juanlufont committed:
Merge branch 'master' into script-transmission

ku1ik commented:
I've used this script and it worked on the first try. Thank you!

juanlufont committed:
Merge branch 'master' into script-transmission

qdm12 commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,121 @@
+fi
+
+if [ "${_USECRED}" ]; then
+    # make sure username AND password were provided
+    if [ -z "${USERNAME}" ] || [ -z "${PASSWORD}" ]; then
+        echo "ERROR: Username or password not provided."
+        exit 1
```
nit you can split this in two if conditions, to have a more precise error message, either username not provided or password not provided

juanlufont replied:
Done

qdm12 commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,121 @@
+    if [ -z "${USERNAME}" ] || [ -z "${PASSWORD}" ]; then
+        echo "ERROR: Username or password not provided."
+        exit 1
+    else
```
nit no need for the else since the if runs `exit 1` (aka early return)

qdm12 commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,121 @@
+if [ "$RES" = "$SUCCESS" ]; then
+    echo "Success! Peer port updated to ${PORT}"
+    exit 0
+else
```
nit no need for the else given the early return with `exit 0` above

also - related to the change proposed - it would be nice to have the error path in an if condition, and the success path at indentation zero below it, so more things can be added after a success in the future without change.

juanlufont replied:
Done

qdm12 commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,121 @@
+#!/bin/sh
```
This script is meant to be run at the port forwarding setup completed event right?
If so, see [this page](https://github.com/qdm12/gluetun-wiki/blob/ee041f335fc0a2e27d4a3c59e37a8957f6dae23f/setup/advanced/vpn-port-forwarding.md#custom-port-forwarding-updown-command)

- we need to use `PORTS` instead of `PORT`, and process possibly multiple ports (or just take the first one from a csv of ports)
- let's copy your script in the Docker image. Add a `COPY extras/scripts/transmission-port-update.sh /somepathofyourchoosing` right below the `HEALTHCHECK ` line
- create a PR to the gluetun wiki on that port forwarding page mentioned above, to tell users about that script they can use - up to you, a new section might be appropriate

Thanks!

juanlufont replied:
Thanks a lot for all the feedback, I will work on the script in the next days :+1:

juanlufont replied:
Transmission expects as peer-port value a single integer. To support `{{PORTS}}` I am splitting by commas (if any) and keeping only the first value.

juanlufont replied:
I have updated the Dockerfile

juanlufont replied:
I created a PR for the Gluetun wiki: https://github.com/qdm12/gluetun-wiki/pull/132

ku1ik commented:
Just reporting that I've been using this for over a month and it's been working flawlessly 👌

juanlufont committed:
Add script improvements

- better error handling
- support for comma-separated list of ports (use only first one)
- style corrections
- improved messages

juanlufont committed:
Add script Transmission script to Dockerfile

juanlufont committed:
Merge branch 'master' into script-transmission

qdm12 commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,132 @@
+    fi
+    # basic auth options, --auth-no-challenge avoids 409 Conflict
+    WGET_OPTS="
+            ${WGET_OPTS}
```
```suggestion
            --quiet
```

qdm12 commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,132 @@
+# check string returned by wget
+SUCCESS='{"arguments":{},"result":"success"}'
+if [ "$RES" != "$SUCCESS" ]; then
+    echo "ERROR: Could not update Transmission peer-port"
```
```suggestion
    echo "ERROR: Could not update Transmission peer-port: ${RES}"
```

qdm12 commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,132 @@
+fi
+
+echo "Success! Transmission peer-port updated to ${PORT}"
+exit 0
```
no need for the final exit 😉 (afaik)

```suggestion
```

qdm12 reviewed (commented):
Thanks for the review of the PR (and the wiki PR 👍 ) Just a few minor things to do if possible 🙏 thanks!

juanlufont committed:
Improve error message

Co-authored-by: Quentin McGaw <quentin.mcgaw@gmail.com>

juanlufont committed:
Remove redundant exit 0

Co-authored-by: Quentin McGaw <quentin.mcgaw@gmail.com>

juanlufont committed:
Move --quiet out of WGET_OPTS

juanlufont commented on extras/scripts/transmission-port-update.sh:
```diff
@@ -0,0 +1,132 @@
+
+# default values
+DEFAULT_URL="http://localhost:9091/transmission/rpc"
+WGET_OPTS="--quiet"
```
I have moved `--quiet` flag out of `WGET_OPTS` but I am keeping the variable to store there, if needed, the API credentials and some extra flags they require. Is this ok?

iridial commented:
I did some tests in k3s with gluetun `v3.40` running as `initContainer` for transmission, and using `VPN_PORT_FORWARDING_UP_COMMAND` as trigger to call the script using: `'/bin/sh -c "/scripts/transmission-port-update.sh --port {{PORTS}}"'`.
The transmission container takes more time to be available to receive the port forward update request, and this leads the initial call of the script to fail.
I suggest to include an initial check if transmission api service is available, and if not, retry once more after configurable timeout with `/bin/sleep <value>`.

sboger commented:
I'm just using this with great success...

`      - VPN_PORT_FORWARDING_UP_COMMAND=/bin/sh -c 'apk add transmission-remote && transmission-remote localhost -p {{PORTS}} && transmission-remote localhost -t all --reannounce'
`

qdm12 commented:
> The transmission container takes more time to be available to receive the port forward update request, and this leads the initial call of the script to fail.

Let's add `--retry-connrefused -t 3` to the wget command
